File: Shared/TandemProtocol/Sources/TandemProtocol/Messages.swift
import Foundation

// JSON control messages. All are additive: unknown keys are ignored by Codable.

public enum Transport: String, Codable, Sendable { case usb, lan }
public enum Platform: String, Codable, Sendable { case ios, android, macos }
public enum Codec: String, Codable, Sendable { case hevc, h264 }
public enum DisplayMode: String, Codable, Sendable { case extend, mirror }

public enum Orientation: String, Codable, Sendable {
    case portrait, landscapeLeft, landscapeRight, portraitUpsideDown
    public var isLandscape: Bool { self == .landscapeLeft || self == .landscapeRight }
}

public struct SafeInsets: Codable, Equatable, Sendable {
    public var top: Double, bottom: Double, left: Double, right: Double
    public init(top: Double = 0, bottom: Double = 0, left: Double = 0, right: Double = 0) {
        self.top = top; self.bottom = bottom; self.left = left; self.right = right
    }
}

public struct ScreenInfo: Codable, Equatable, Sendable {
    /// Usable full-screen size in physical pixels in the *current* orientation.
    public var wPx: Int
    public var hPx: Int
    public var scale: Double
    public var refreshHz: Double
    public var safeInsets: SafeInsets?
    public init(wPx: Int, hPx: Int, scale: Double, refreshHz: Double, safeInsets: SafeInsets? = nil) {
        self.wPx = wPx; self.hPx = hPx; self.scale = scale; self.refreshHz = refreshHz; self.safeInsets = safeInsets
    }
}

public struct AppInfo: Codable, Equatable, Sendable {
    public var name: String
    public var version: String
    public var build: Int
    public init(name: String, version: String, build: Int) {
        self.name = name; self.version = version; self.build = build
    }
}

public struct Hello: Codable, Equatable, Sendable {
    public var pv: Int
    public var hostId: String
    public var hostName: String
    public var transport: Transport
    public var hostNonce: Data          // base64 in JSON
    public var encrypt: Bool
    public var app: AppInfo?
    public init(pv: Int = Tandem.protocolVersion, hostId: String, hostName: String, transport: Transport,
                hostNonce: Data, encrypt: Bool, app: AppInfo? = nil) {
        self.pv = pv; self.hostId = hostId; self.hostName = hostName; self.transport = transport
        self.hostNonce = hostNonce; self.encrypt = encrypt; self.app = app
    }
}

public struct HelloAck: Codable, Equatable, Sendable {
    public var pv: Int
    public var deviceId: String
    public var deviceName: String
    public var platform: Platform
    public var model: String?
    public var deviceNonce: Data
    public var screen: ScreenInfo
    public var orientation: Orientation
    public var codecs: [Codec]
    public var maxDecode: Size?
    public var features: [String]
    public var paired: Bool
    public var authRequired: Bool
    public init(pv: Int = Tandem.protocolVersion, deviceId: String, deviceName: String, platform: Platform,
                model: String? = nil, deviceNonce: Data, screen: ScreenInfo, orientation: Orientation,
                codecs: [Codec], maxDecode: Size? = nil, features: [String], paired: Bool, authRequired: Bool) {
        self.pv = pv; self.deviceId = deviceId; self.deviceName = deviceName; self.platform = platform
        self.model = model; self.deviceNonce = deviceNonce; self.screen = screen; self.orientation = orientation
        self.codecs = codecs; self.maxDecode = maxDecode; self.features = features
        self.paired = paired; self.authRequired = authRequired
    }
}

public struct Size: Codable, Equatable, Sendable {
    public var w: Int, h: Int
    public init(w: Int, h: Int) { self.w = w; self.h = h }
}

public struct PairRequired: Codable, Equatable, Sendable {
    public var attemptsLeft: Int
    public init(attemptsLeft: Int) { self.attemptsLeft = attemptsLeft }
}
public struct PairMessage: Codable, Equatable, Sendable {
    public var proof: Data
    public init(proof: Data) { self.proof = proof }
}
public struct PairOK: Codable, Equatable, Sendable {
    public var box: Data
    public init(box: Data) { self.box = box }
}
public struct PairFail: Codable, Equatable, Sendable {
    public var attemptsLeft: Int
    public var locked: Bool
    public init(attemptsLeft: Int, locked: Bool) { self.attemptsLeft = attemptsLeft; self.locked = locked }
}
public struct AuthMessage: Codable, Equatable, Sendable {
    public var proof: Data
    public init(proof: Data) { self.proof = proof }
}
public struct AuthOK: Codable, Equatable, Sendable {
    public var proof: Data
    public init(proof: Data) { self.proof = proof }
}
public struct AuthFail: Codable, Equatable, Sendable {
    public var reason: String
    public init(reason: String) { self.reason = reason }
}

public struct StreamConfig: Codable, Equatable, Sendable {
    public var codec: Codec
    public var wPx: Int
    public var hPx: Int
    public var fps: Int
    public var desktop: Size
    public var orientation: Orientation
    public var mode: DisplayMode
    public var cursorBaked: Bool
    public init(codec: Codec, wPx: Int, hPx: Int, fps: Int, desktop: Size, orientation: Orientation,
                mode: DisplayMode = .extend, cursorBaked: Bool = true) {
        self.codec = codec; self.wPx = wPx; self.hPx = hPx; self.fps = fps; self.desktop = desktop
        self.orientation = orientation; self.mode = mode; self.cursorBaked = cursorBaked
    }
}

public struct DeviceConfig: Codable, Equatable, Sendable {
    public var orientation: Orientation
    public var screen: ScreenInfo
    public init(orientation: Orientation, screen: ScreenInfo) { self.orientation = orientation; self.screen = screen }
}

public struct KeyEvent: Codable, Equatable, Sendable {
    public var down: Bool
    public var keyCode: Int
    public var text: String?
    public var mods: [String]
    public init(down: Bool, keyCode: Int, text: String?, mods: [String] = []) {
        self.down = down; self.keyCode = keyCode; self.text = text; self.mods = mods
    }
}

public struct Stats: Codable, Equatable, Sendable {
    public var fpsDecoded: Double
    public var fpsDropped: Double
    public var decodeMsP50: Double
    public var queued: Int
    public var rttMs: Double
    public var e2eMsP50: Double?
    public init(fpsDecoded: Double, fpsDropped: Double, decodeMsP50: Double, queued: Int, rttMs: Double, e2eMsP50: Double? = nil) {
        self.fpsDecoded = fpsDecoded; self.fpsDropped = fpsDropped; self.decodeMsP50 = decodeMsP50
        self.queued = queued; self.rttMs = rttMs; self.e2eMsP50 = e2eMsP50
    }
}

public struct Bye: Codable, Equatable, Sendable {
    public var reason: String
    public init(reason: String) { self.reason = reason }
}

// MARK: - JSON coding

public enum JSONCoding {
    public static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.dataEncodingStrategy = .base64
        e.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        return e
    }()
    public static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.dataDecodingStrategy = .base64
        return d
    }()
}

public extension Frame {
    /// Build a JSON frame.
    static func json<T: Encodable>(_ type: FrameType, _ value: T) throws -> Frame {
        Frame(type: type, payload: try JSONCoding.encoder.encode(value))
    }
    /// Decode a JSON payload.
    func decode<T: Decodable>(_ t: T.Type = T.self) throws -> T {
        try JSONCoding.decoder.decode(T.self, from: payload)
    }
}

File: Receivers/Android/app/src/main/java/app/tandem/receiver/net/Messages.kt
package app.tandem.receiver.net

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

/** Safe-area insets in physical pixels (PROTOCOL.md §3.2). */
data class SafeInsets(val top: Int, val bottom: Int, val left: Int, val right: Int) {
    fun toJson(): JSONObject =
        JSONObject()
            .put("top", top)
            .put("bottom", bottom)
            .put("left", left)
            .put("right", right)

    companion object {
        val ZERO = SafeInsets(0, 0, 0, 0)
    }
}

/** The usable full-screen area, in physical pixels. */
data class ScreenInfo(
    val wPx: Int,
    val hPx: Int,
    val scale: Float,
    val refreshHz: Int,
    val safeInsets: SafeInsets,
) {
    fun toJson(): JSONObject =
        JSONObject()
            .put("wPx", wPx)
            .put("hPx", hPx)
            .put("scale", scale.toDouble())
            .put("refreshHz", refreshHz)
            .put("safeInsets", safeInsets.toJson())
}

/** PROTOCOL.md §3.2 `orientation`. */
enum class Orientation(val wire: String) {
    PORTRAIT("portrait"),
    LANDSCAPE_LEFT("landscapeLeft"),
    LANDSCAPE_RIGHT("landscapeRight"),
    PORTRAIT_UPSIDE_DOWN("portraitUpsideDown"),
    ;

    companion object {
        fun fromWire(value: String?): Orientation =
            entries.firstOrNull { it.wire == value } ?: PORTRAIT
    }
}

/** Everything the receiver advertises about itself in HELLO_ACK. */
data class DeviceProfile(
    val deviceId: String,
    val deviceName: String,
    val model: String,
    val screen: ScreenInfo,
    val orientation: Orientation,
    val codecs: List<String>,
    val maxDecodeWidth: Int,
    val maxDecodeHeight: Int,
    val features: List<String>,
)

/** HELLO, host to receiver (PROTOCOL.md §3.1). */
data class Hello(
    val pv: Int,
    val hostId: String,
    val hostName: String,
    val transport: String,
    val hostNonce: ByteArray,
    val encrypt: Boolean,
    val appName: String,
    val appVersion: String,
    val appBuild: Int,
) {
    val isUsb: Boolean get() = transport.equals("usb", ignoreCase = true)

    override fun equals(other: Any?): Boolean =
        this === other ||
            (
                other is Hello && pv == other.pv && hostId == other.hostId &&
                    hostName == other.hostName && transport == other.transport &&
                    hostNonce.contentEquals(other.hostNonce) && encrypt == other.encrypt
            )

    override fun hashCode(): Int = hostId.hashCode() * 31 + hostNonce.contentHashCode()

    companion object {
        fun parse(payload: ByteArray): Hello {
            val json = Json.obj(payload)
            val nonce = Json.base64(json, "hostNonce")
            if (nonce.size != 16) throw ProtocolException("hostNonce must be 16 bytes, got ${nonce.size}")
            val app = json.optJSONObject("app")
            return Hello(
                pv = json.optInt("pv", 0),
                hostId = json.optString("hostId", ""),
                hostName = json.optString("hostName", ""),
                transport = json.optString("transport", "lan"),
                hostNonce = nonce,
                encrypt = json.optBoolean("encrypt", false),
                appName = app?.optString("name", "") ?: "",
                appVersion = app?.optString("version", "") ?: "",
                appBuild = app?.optInt("build", 0) ?: 0,
            ).also {
                if (it.hostId.isEmpty()) throw ProtocolException("HELLO is missing hostId")
            }
        }
    }
}

/** STREAM_CONFIG, host to receiver (PROTOCOL.md §3.4). */
data class StreamConfig(
    val codec: String,
    val wPx: Int,
    val hPx: Int,
    val fps: Int,
    val desktopW: Int,
    val desktopH: Int,
    val orientation: Orientation,
    val mode: String,
    val cursorBaked: Boolean,
) {
    /** A decoder reset is required only when these three change (§3.4). */
    fun decoderIdentity(): Triple<String, Int, Int> = Triple(codec.lowercase(), wPx, hPx)

    companion object {
        fun parse(payload: ByteArray): StreamConfig {
            val json = Json.obj(payload)
            val desktop = json.optJSONObject("desktop")
            val w = json.optInt("wPx", 0)
            val h = json.optInt("hPx", 0)
            if (w <= 0 || h <= 0) throw ProtocolException("STREAM_CONFIG has a non-positive frame size")
            return StreamConfig(
                codec = json.optString("codec", "hevc"),
                wPx = w,
                hPx = h,
                fps = json.optInt("fps", 60),
                desktopW = desktop?.optInt("w", 0) ?: 0,
                desktopH = desktop?.optInt("h", 0) ?: 0,
                orientation = Orientation.fromWire(json.optString("orientation")),
                mode = json.optString("mode", "extend"),
                cursorBaked = json.optBoolean("cursorBaked", true),
            )
        }
    }
}

/** Receiver-side view of the decoded STATS payload (PROTOCOL.md §7). */
data class StatsSnapshot(
    val fpsDecoded: Double,
    val fpsDropped: Double,
    val decodeMsP50: Double,
    val queued: Int,
    val rttMs: Double,
    val e2eMsP50: Double,
) {
    fun toJsonBytes(): ByteArray =
        JSONObject()
            .put("fpsDecoded", round1(fpsDecoded))
            .put("fpsDropped", round1(fpsDropped))
            .put("decodeMsP50", round1(decodeMsP50))
            .put("queued", queued)
            .put("rttMs", round1(rttMs))
            .put("e2eMsP50", round1(e2eMsP50))
            .toString()
            .toByteArray(Charsets.UTF_8)

    private fun round1(v: Double): Double = Math.round(v * 10.0) / 10.0
}

/** KEY, receiver to host (PROTOCOL.md §6.3). */
data class KeyEvent(
    val down: Boolean,
    val keyCode: Int,
    val text: String,
    val mods: List<String>,
) {
    fun toJsonBytes(): ByteArray {
        val modsArray = JSONArray()
        mods.forEach { modsArray.put(it) }
        return JSONObject()
            .put("down", down)
            .put("keyCode", keyCode)
            .put("text", text)
            .put("mods", modsArray)
            .toString()
            .toByteArray(Charsets.UTF_8)
    }
}

/** JSON helpers shared by every message codec. */
object Json {
    fun obj(payload: ByteArray): JSONObject =
        try {
            JSONObject(String(payload, Charsets.UTF_8))
        } catch (e: JSONException) {
            throw ProtocolException("payload is not a JSON object", e)
        }

    fun bytes(json: JSONObject): ByteArray = json.toString().toByteArray(Charsets.UTF_8)

    fun base64(json: JSONObject, key: String): ByteArray {
        val value = json.optString(key, "")
        if (value.isEmpty()) throw ProtocolException("missing base64 field '$key'")
        return try {
            java.util.Base64.getDecoder().decode(value)
        } catch (e: IllegalArgumentException) {
            throw ProtocolException("field '$key' is not valid base64", e)
        }
    }

    // java.util.Base64 rather than android.util.Base64: it exists from API 26
    // (this app's floor) and, unlike the Android one, it is not a stub in JVM
    // unit tests, so the handshake codecs are covered without a device.
    fun encodeBase64(bytes: ByteArray): String = java.util.Base64.getEncoder().encodeToString(bytes)
}
